### PR TITLE
restricted metadata changes

### PR DESCRIPTION
**Pull request by Karthikmvdp**

### CODE REVIEW

**Comment by chhagarw** [comment]
@Karthikmvdp can u also add the changes to have volume metadata,  model_update to have vdisk_uid assigned as volume_wwn.

 model_update = super(PowerPureFCDriver, self).\
            create_volume(vol_copy)
        return self._model_update(model_update, vol_copy)

    def _model_update(self, model_update, volume):
        """add volume wwn to the metadata of the new volume"""
        if not model_update:
            model_update = {}
        meta = self.db.volume_metadata_get(
            context.get_admin_context(), volume['id'])
        model_update['metadata'] = meta if meta else dict()
        attrs = self.get_volume_info(volume)
        model_update['metadata']['volume_wwn'] = attrs['vdisk_UID']
        return model_update

File: pure_powervc.py [hunk 1]
```diff
@@ -33,9 +33,43 @@
...
+                with excutils.save_and_reraise_exception():
+                    LOG.info(_("Create restricted metadata failed. "
+                               "Module: %(f_module)s, "
+                               "Function: %(f_name)s, "
+                               "Volume: %(volume)s, "
+                               "Error: %(err)s" %
+                               dict(f_module=f.__module__,
+                                    f_name=f.__name__,
+                                    volume=args[vol_arg]['id'],
+                                    err=six.text_type(ex))))
```
Comment (sdodsley) [inline pure_powervc.py]: `six` is no longer to be used in Cinder code - please change.

File: pure_powervc.py [hunk 2]
```diff
@@ -33,9 +33,43 @@
...
 from powervc_cinder.volume.discovery_driver import PORT_STATUS
 from powervc_cinder.volume.discovery_driver import UNKNOWN_VALUE
+from powervc_cinder.db import api as powervc_db_api
+from oslo_utils import excutils, importutils
+from cinder import exception
+from powervc_cinder import exception as powervc_exception
+from oslo_utils import units
+
+import json
+import six
```
Comment (sdodsley) [inline pure_powervc.py]: `six` not approved in latest Cinder code base

**Review by sdodsley — CHANGES_REQUESTED** [review]

**Review by chhagarw — COMMENTED** [review]
@sdodsley currently we are using get_volume_info, which uses  
pure_volumes = current_array.list_volumes()
and traverse to get the volume details, does purestorage has an interface to get the volume info of a single volume

**Comment by sdodsley** [comment]
Yes - I sent details over - you can use `get_volume(<volume_name)`

File: pure_powervc.py [hunk 3]
```diff
@@ -33,9 +33,43 @@
...
+RESTRICTED_METADATA_VDISK_UID_KEY = "vdisk_uid"
+RESTRICTED_METADATA_VDISK_NAME_KEY = "vdisk_name"
 
 LOG = logging.getLogger(__name__)
 
+def create_restricted_metadata(vol_arg=1):
+    """add restricted metadata for a new volume"""
+    def wrap(f):
+        def decorator(*args, **kwargs):
+            args[0]._validate_type(args[vol_arg])
```
Comment (sdodsley) [inline pure_powervc.py]: I don't see a function definition for `_validate_type` so this will cause the driver to crash

**Review by sdodsley — CHANGES_REQUESTED** [review]

**Comment by sdodsley** [comment]
@Karthikmvdp are you sure this is correct?
I just used this latest version in my PowerVC 2.0.0 system and it crashed with the following error
```
2021-04-27 12:37:05.518 1764570 ERROR cinder Traceback (most recent call last):
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/bin/cinder-volume", line 10, in <module>
2021-04-27 12:37:05.518 1764570 ERROR cinder     sys.exit(main())
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/cinder/cmd/volume.py", line 232, in main
2021-04-27 12:37:05.518 1764570 ERROR cinder     _launch_services_pvc()
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/cinder/cmd/volume.py", line 194, in _launch_services_pvc
2021-04-27 12:37:05.518 1764570 ERROR cinder     _launch_service_pvc(launcher, CONF.backend_host)
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/cinder/cmd/volume.py", line 128, in _launch_service_pvc
2021-04-27 12:37:05.518 1764570 ERROR cinder     cluster=CONF.cluster)
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/cinder/service.py", line 406, in create
2021-04-27 12:37:05.518 1764570 ERROR cinder     cluster=cluster, **kwargs)
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/cinder/service.py", line 156, in __init__
2021-04-27 12:37:05.518 1764570 ERROR cinder     *args, **kwargs)
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/powervc_cinder/volume/manager.py", line 494, in __init__
2021-04-27 12:37:05.518 1764570 ERROR cinder     service_name=service_name, *args, **kwargs)
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/cinder/volume/manager.py", line 275, in __init__
2021-04-27 12:37:05.518 1764570 ERROR cinder     active_backend_id=curr_active_backend_id)
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/oslo_utils/importutils.py", line 44, in import_object
2021-04-27 12:37:05.518 1764570 ERROR cinder     return import_class(import_str)(*args, **kwargs)
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/oslo_utils/importutils.py", line 30, in import_class
2021-04-27 12:37:05.518 1764570 ERROR cinder     __import__(mod_str)
2021-04-27 12:37:05.518 1764570 ERROR cinder   File "/usr/lib/python3.6/site-packages/cinder/volume/drivers/pure_powervc.py", line 46, in <module>
2021-04-27 12:37:05.518 1764570 ERROR cinder     from powervc_cinder.volume.discovery_driver \
2021-04-27 12:37:05.518 1764570 ERROR cinder ImportError: cannot import name 'RESTRICTED_METADATA_VDISK_pg83NAA_KEY'

```

**Comment by Karthikmvdp** [comment]
@chhagarw @shyamav @sdodsley I have updated my change set.

**Comment by sdodsley** [comment]
@Karthikmvdp do you have an updated `discovery_driver.py` I can test this with, or will upgrading to 2.0.1 resolve this crash?

**Comment by Karthikmvdp** [comment]
@sdodsley we have changes in discovery driver, you need to update the discovery_driver.py to resolve the crash.

**Comment by sdodsley** [comment]
@Karthikmvdp can you send me the updated version so I can test out the fixes?

**Comment by Karthikmvdp** [comment]
@sdodsley sent changes via mail.

**Comment by sdodsley** [comment]
OK - I can't test LKU in my environment, but it runs fine with those updated files.

**Pull request merged**